### ISSUE TITLE
torch.std NaN gradient

### ISSUE THREAD

**Issue by elbaro** [pytorch/pytorch#4320]
`torch.std(x)` gives NaN gradients at `x=0`.

```py
import torch
from torch.autograd.variable import Variable
import torch.nn.functional as F

a = Variable(torch.FloatTensor(16,3,64,64).zero_(),requires_grad=True)

# b = a.std(dim=0).mean()  # NaN gradients
# b = (a - a.mean(dim=0,keepdim=True)).norm(p=2,dim=0).mean() / (15**0.5)  # 0 gradients

b.backward()

print(a.grad.data)
```

related #2421 #3264

cc @ezyang @SsnL @albanD @zou3519 @gqchen @gchanan

**Comment by rwebb** [comment]
You also get NaN for any constant input not just zeros.  Replace zeros_ with ones_ to see that.

**Comment by soumith** [comment]
@albanD what do you want this to be (if not nan), or do you want to keep it a nan?

**Comment by albanD** [comment]
This is the exact same thing as the norm: there is no gradient at when the output is `0` (all points are equal) and a tensor full of `0`s is a valid subgradient at that point.
There are the same pros and cons as for the norm case.
For consistency reason I think we should return a Tensor of `0`s.

**Comment by ludwigwinkler** [comment]
Somewhat related:
When we apply torch.std( arbitrary_tensor,  dim= dimension_with_single_entry ) and the dimension has only one entry, then nan is returned.

'''
a = torch.randn(1,10,10)
std = torch.std(a, dim=0)
print(std) -> is a 10x10 tensor of nan's
''' 

Maybe a warning could be raised with regard to the nan computation that all subsequent computations with the standard deviation will result in nan's?

I'm aware of the unbiased estimator which returns 0 in this case.
Nevertheless it would be nice to catch the nan's.

**Comment by VitalyFedyunin** [comment]
Might extend to other functions, requires checking.

**Comment by albanD** [comment]
Hi,
No one had a chance to get around this and update the formula I'm afraid. But we would be happy to accep a PR doing it though!
We want to add a line similar to [this one](https://github.com/vishwakftw/pytorch/blob/ede9bc97c3d734f3c80f4c0c08e1fe3dc2ab0250/tools/autograd/templates/Functions.cpp#L134-L135) to the std formula
